### PR TITLE
[VitisAI] custom op support multiple outputs

### PR DESCRIPTION
**Pull request by mingyueliuh**
### Description
The implementation inside EP requires registering some custom ops which are only used in the model compilation phase. Currently only single output is supported.

### Motivation and Context
Now the demand upgrade requires support for multiple outputs, so the shaper infer of ep custom op needs to be extended to support multiple outputs

### CODE REVIEW

**Comment by mingyueliuh** [comment]
@jywu-msft please take a look when you are available. Thanks.

**Comment by jywu-msft** [comment]
can you provide some more details on why are the custom ops needed? if they are only needed during model compilation, can they be completely self contained within vitis ai ep implementation (rather than needing to add static functions in provider_bridge?)

File: onnxruntime/core/session/provider_bridge_ort.cc [hunk 1]
```diff
@@ -649,17 +647,43 @@ struct ProviderHostImpl : ProviderHost {
...
+    return elemType;
+  }
+
+  static void xir_shape_infer(ONNX_NAMESPACE::InferenceContext& ctx) {
+    auto num_output = ctx.getNumOutputs();
+    if (num_output == 1) {
+      auto* shape = ctx.getAttribute("shape");
+      auto* data_type = ctx.getAttribute("data_type");
+      if (data_type == nullptr) {
+        std::cerr << "[VitisAI] custom op is missing `data_type` attr." << std::endl;
```
Comment (HectorSVC) [inline onnxruntime/core/session/provider_bridge_ort.cc:660]: this shouldn't limit to VitisAI EP.

**Comment by HectorSVC** [comment]
Is it possible to enable UT to cover this code?

File: onnxruntime/core/session/provider_bridge_ort.cc (hunk 1 above)
Comment (mingyueliuh) [reply onnxruntime/core/session/provider_bridge_ort.cc]: Sure , Thanks for your review.

**Comment by mingyueliuh** [comment]
> can you provide some more details on why are the custom ops needed? if they are only needed during model compilation, can they be completely self contained within vitis ai ep implementation (rather than needing to add static functions in provider_bridge?)

Hi @HectorSVC ， During the model compilation phase, the EP will first convert the ONNX model into the Vitis AI custom IR format. Due to some differences between the two IRs, the graph exists in an intermediate state during the conversion. We use an ONNX Runtime custom op to represent this intermediate model and rely on the functionalities provided by ONNX Runtime to ensure the integrity and validity of the graph.

**Comment by mingyueliuh** [comment]
> Is it possible to enable UT to cover this code?



> Is it possible to enable UT to cover this code?
Hi @HectorSVC ， Since ONNXRuntime is not compiled with Vitis AI EP by default (without --use_vitisai), I think we cannot enable UT for cover this.

File: onnxruntime/core/session/provider_bridge_ort.cc [hunk 2]
```diff
@@ -649,17 +647,43 @@ struct ProviderHostImpl : ProviderHost {
...
+        }
+      } else {
+        // set scalar type.
+        ONNX_NAMESPACE::getOutputShape(ctx, 0, ONNX_NAMESPACE::TypeProto::kTensorType)->clear_dim();
       }
     } else {
-      // set scalar type.
-      ONNX_NAMESPACE::getOutputShape(ctx, 0, ONNX_NAMESPACE::TypeProto::kTensorType)->clear_dim();
+      for (auto idx = 0u; idx < num_output; idx++) {
+        auto* shape = ctx.getAttribute("shape_" + idx);
```
Comment (HectorSVC) [inline onnxruntime/core/session/provider_bridge_ort.cc:675]: idx is int not string. it causes the build error:
/mnt/vss/_work/1/s/onnxruntime/core/session/provider_bridge_ort.cc:676:49: error: adding 'unsigned int' to a string does not append to the string [-Werror,-Wstring-plus-int]
        auto* shape = ctx.getAttribute("shape_" + idx);
                                       ~~~~~~~~~^~~~~
/mnt/vss/_work/1/s/onnxruntime/core/session/provider_bridge_ort.cc:676:49: note: use array indexing to silence this warning
        auto* shape = ctx.getAttribute("shape_" + idx);
                                                ^
                                       &        [    ]
/mnt/vss/_work/1/s/onnxruntime/core/session/provider_bridge_ort.cc:677:57: error: adding 'unsigned int' to a string does not append to the string [-Werror,-Wstring-plus-int]
        auto* data_type = ctx.getAttribute("data_type_" + idx);
                                           ~~~~~~~~~~~~~^~~~~
/mnt/vss/_work/1/s/onnxruntime/core/session/provider_bridge_ort.cc:677:57: note: use array indexing to silence this warning
        auto* data_type = ctx.getAttribute("data_type_" + idx);
                                                        ^
                                           &            [    ]
2 errors generated.

Comment (mingyueliuh) [reply onnxruntime/core/session/provider_bridge_ort.cc]: Thanks  @HectorSVC ， this build error has been fixed.

**Review by HectorSVC — APPROVED** [review]

**Pull request merged**